File: leetcode/my-calendar-i.cpp
/* coding:utf-8
 * Copyright (C) dirlt
 */

#include <algorithm>
#include <map>
#include <queue>
#include <set>
#include <string>
#include <vector>
using namespace std;

class MyCalendar {
   public:
    map<int, int> counter;
    MyCalendar() {}

    bool book(int start, int end) {
        counter[start] += 1;
        counter[end] -= 1;
        int overlap = 0;
        auto ans = true;
        for (auto it = counter.begin(); it != counter.end(); ++it) {
            if (it->first > end) {
                break;
            }
            overlap += it->second;
            if (overlap >= 2) {
                ans = false;
                break;
            }
        }
        if (!ans) {
            counter[start] -= 1;
            counter[end] += 1;
        }
        return ans;
    }
};

File: leetcode/my-calendar-i.py
#!/usr/bin/env python
# coding:utf-8
# Copyright (C) dirlt
import bisect


class MyCalendar:

    def __init__(self):
        self.xs = []
        self.ys = []

    def book(self, start, end):
        """
        :type start: int
        :type end: int
        :rtype: bool
        """
        i = bisect.bisect_left(self.xs, start)
        # compare to self.ys[i-1] and self.xs[i]
        if (i - 1) >= 0:
            if start < self.ys[i - 1]:
                return False
        if i < len(self.xs):
            if end > self.xs[i]:
                return False
        self.xs.insert(i, start)
        self.ys.insert(i, end)
        return True

# Your MyCalendar object will be instantiated and called as such:
# obj = MyCalendar()
# param_1 = obj.book(start,end)
